da/da_test.go

```diff
@@ -0,0 +1,124 @@
+package da
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/rollkit/go-da/proxy"
+	goDATest "github.com/rollkit/go-da/test"
+	"github.com/rollkit/rollkit/types"
+)
+
+const mockDaBlockTime = 100 * time.Millisecond
+
+func TestMain(m *testing.M) {
+	srv := startMockGRPCServ()
+	if srv == nil {
+		os.Exit(1)
+	}
+	exitCode := m.Run()
+
+	// teardown servers
+	srv.GracefulStop()
+
+	os.Exit(exitCode)
+}
+
+func TestSubmitRetrieve(t *testing.T) {
+	dummyClient := &DAClient{DA: goDATest.NewDummyDA(), Logger: log.TestingLogger()}
+	grpcClient, err := startMockGRPCClient()
+	require.NoError(t, err)
+	clients := map[string]*DAClient{
+		"dummy": dummyClient,
+		"grpc":  grpcClient,
+	}
+	for name, dalc := range clients {
+		t.Run(name, func(t *testing.T) {
+			doTestSubmitRetrieve(t, dalc)
+		})
+	}
+}
+
+func startMockGRPCServ() *grpc.Server {
+	srv := proxy.NewServer(goDATest.NewDummyDA(), grpc.Creds(insecure.NewCredentials()))
+	lis, err := net.Listen("tcp", "127.0.0.1"+":"+strconv.Itoa(7980))
+	if err != nil {
+		fmt.Println(err)
+		return nil
+	}
+	go func() {
+		_ = srv.Serve(lis)
+	}()
+	return srv
+}
+
+func startMockGRPCClient() (*DAClient, error) {
+	client := proxy.NewClient()
+	err := client.Start("127.0.0.1:7980", grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	return &DAClient{DA: client, Logger: log.TestingLogger()}, nil
+}
+
+func doTestSubmitRetrieve(t *testing.T, dalc *DAClient) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	require := require.New(t)
+	assert := assert.New(t)
+
+	// wait a bit more than mockDaBlockTime, so mock can "produce" some blocks
+	time.Sleep(mockDaBlockTime + 20*time.Millisecond)
+
+	countAtHeight := make(map[uint64]int)
+	blockToDAHeight := make(map[*types.Block]uint64)
+	numBatches := uint64(10)
+	blocksSubmittedPerBatch := 10
+
+	for i := uint64(0); i < numBatches; i++ {
+		blocks := make([]*types.Block, blocksSubmittedPerBatch)
+		for j := 0; j < len(blocks); j++ {
+			blocks[j] = types.GetRandomBlock(i*numBatches+uint64(j), rand.Int()%20) //nolint:gosec
+		}
+		resp := dalc.SubmitBlocks(ctx, blocks)
+		assert.Equal(StatusSuccess, resp.Code, resp.Message)
+		time.Sleep(time.Duration(rand.Int63() % mockDaBlockTime.Milliseconds())) //nolint:gosec
+
+		for _, b := range blocks {
+			blockToDAHeight[b] = resp.DAHeight
+			countAtHeight[resp.DAHeight]++
+		}
+	}
+
+	// wait a bit more than mockDaBlockTime, so mock can "produce" last blocks
+	time.Sleep(mockDaBlockTime + 20*time.Millisecond)
```

Comment:
[follow up] waiting for n number of blocks to be produced is a pretty common thing. We should maybe think of how to refactor this into a helper method from go-da